### PR TITLE
Fixes to runtime.Poller

### PR DESCRIPTION
**Pull request by jhendrixMSFT**
Fixed the behavior of Poll and Result methods and updated their doc
comments explaining the behavior.
Clarified some other Poller doc comments and improved error messages.

- [ ] The purpose of this PR is explained in this or a referenced issue.
- [ ] The PR does not update generated files.
   - These files are managed by the codegen framework at [Azure/autorest.go][].
- [ ] Tests are included and/or updated for code changes.
- [ ] Updates to [CHANGELOG.md][] are included.
- [ ] MIT license headers are included in each file.

[Azure/autorest.go]: https://github.com/Azure/autorest.go
[CHANGELOG.md]: https://github.com/Azure/azure-sdk-for-go/blob/main/CHANGELOG.md

### CODE REVIEW

File: sdk/azcore/runtime/poller_test.go [hunk 1]
```diff
@@ -380,6 +391,71 @@ func TestOpPollerWithWidgetPUT(t *testing.T) {
...
+		Request: &http.Request{
+			Method: http.MethodPut,
+			URL:    reqURL,
+		},
+	}
+	pl := newTestPipeline(&policy.ClientOptions{Transport: srv})
+	lro, err := NewPoller[widget](firstResp, pl, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (chlowell) [inline sdk/azcore/runtime/poller_test.go:436]:
```suggestion
	require.Nil(t, err)
```

File: sdk/azcore/runtime/poller.go [hunk 2]
```diff
@@ -250,46 +252,69 @@ func (p *Poller[T]) PollUntilDone(ctx context.Context, options *PollUntilDoneOpt
...
+	if p.done {
+		// the result has already been retrieved, return the cached value
+		if p.err != nil {
+			return *new(T), p.err
+		}
+		return *p.result, nil
+	}
+	res, err := p.op.Result(ctx, p.result)
+	var respErr *exported.ResponseError
+	if err != nil && errors.As(err, &respErr) {
```
Comment (chlowell) [inline sdk/azcore/runtime/poller.go:294]: Won't `errors.As` return false when `err == nil`?

File: sdk/azcore/runtime/poller_test.go [hunk 3]
```diff
@@ -380,6 +391,71 @@ func TestOpPollerWithWidgetPUT(t *testing.T) {
...
+	// prevent the retry policy from masking this transient error
+}
+
+func TestOpPollerWithWidgetFinalGetError(t *testing.T) {
+	srv, close := mock.NewServer()
+	srv.AppendResponse(mock.WithStatusCode(http.StatusAccepted), mock.WithBody([]byte(`{"status": "InProgress"}`)))
+	srv.AppendResponse(mock.WithStatusCode(http.StatusOK), mock.WithBody([]byte(`{"status": "Succeeded"}`)))
+	// the first attempt at a final GET returns an error
+	srv.AppendError(&nonRetriableError{Msg: "failed attempt"})
+	// PUT and PATCH state that a final GET will happen
```
Comment (chlowell) [inline sdk/azcore/runtime/poller_test.go:412]: Nitpicking: seems to belong before the prior comment

**Review by chlowell — APPROVED** [review]

File: sdk/azcore/runtime/poller_test.go (hunk 1 above)
Comment (jhendrixMSFT) [reply sdk/azcore/runtime/poller_test.go]: Copy/pasta, fixed

File: sdk/azcore/runtime/poller.go (hunk 2 above)
Comment (jhendrixMSFT) [reply sdk/azcore/runtime/poller.go]: Yep good catch, fixed.

File: sdk/azcore/runtime/poller_test.go (hunk 3 above)
Comment (jhendrixMSFT) [reply sdk/azcore/runtime/poller_test.go]: Fixed

**Pull request merged**